packages/survey-core/src/question_baseselect.ts

```diff
@@ -1931,40 +1931,48 @@ export class QuestionSelectBase extends Question {
     return settings.itemFlowDirection;
   }
   get columns() {
-    var columns = [];
-    var colCount = this.getCurrentColCount();
-    if (this.hasColumns && this.renderedChoices.length > 0) {
-      let choicesToBuildColumns = (!this.separateSpecialChoices && !this.isInDesignMode) ?
-        this.renderedChoices : this.dataChoices;
-      if (this.itemFlowDirection === "column") {
-        var prevIndex = 0;
-        var leftElementsCount = choicesToBuildColumns.length % colCount;
-        for (var i = 0; i < colCount; i++) {
-          var column = [];
-          for (
-            var j = prevIndex;
-            j < prevIndex + Math.floor(choicesToBuildColumns.length / colCount);
-            j++
-          ) {
-            column.push(choicesToBuildColumns[j]);
-          }
-          if (leftElementsCount > 0) {
-            leftElementsCount--;
-            column.push(choicesToBuildColumns[j]);
-            j++;
-          }
-          prevIndex = j;
-          columns.push(column);
-        }
-      } else {
-        for (var i = 0; i < colCount; i++) {
-          var column = [];
-          for (var j = i; j < choicesToBuildColumns.length; j += colCount) {
-            column.push(choicesToBuildColumns[j]);
-          }
-          columns.push(column);
+    if (!this.hasColumns || this.renderedChoices.length === 0) return [];
+
+    const colCount = this.getCurrentColCount();
+    let choicesToBuildColumns = (!this.separateSpecialChoices && !this.isInDesignMode) ?
+      this.renderedChoices : this.dataChoices;
+
+    if (this.itemFlowDirection === "column") {
+      return this.getColumnsWithColumnItemFlow(choicesToBuildColumns, colCount);
+    } else {
+      return this.getColumnsWithRowItemFlow(choicesToBuildColumns, colCount);
+    }
+  }
+  getColumnsWithColumnItemFlow(choices, colCount) {
+    const columns =[];
+    let maxColumnHeight = Math.floor(choices.length / colCount);
+
+    if (choices.length % colCount) {
+      maxColumnHeight += 1;
+    }
+
+    for (let i = 0; i < colCount; i++) {
+      const column = [];
+      const choiceStartIndex = i * maxColumnHeight;
+      for (let j = 0; j < maxColumnHeight; j++) {
+        const choice = choices[j + choiceStartIndex];
+        if (choice) {
+          column.push(choice);
         }
       }
+      columns.push(column);
+    }
+
+    return columns;
+  }
+  getColumnsWithRowItemFlow(choices, colCount) {
```

Comment:
If we mark this method as private, we'll be free to rename/remove it in the future.